nixos/roles/router/kresd/default.nix

```diff
@@ -0,0 +1,167 @@
+      -- timeout than recommended in the rfc to ensure that clients wait for
+      -- long enough to trigger the stale serving behaviour. we match the
+      -- rfc for the stale serving timeout though so we get the benefits
+      -- if the granularity changes in the future.
+      serve_stale.timeout = 1800
+
+      -- if a backend is detected as down, let it stay down for a minute so we
+      -- can immediately respond to requests. however, this also means backends
+      -- may take a minute to recover
+      cache.ns_tout(60000)
+
+      -- access control: only answer queries from ourselves and own address
+      -- space
+      view:addr('127.0.0.0/8', policy.all(policy.PASS))
+      view:addr('::1/128', policy.all(policy.PASS))
+
+      ${lib.concatMapStringsSep "\n" (
+        net: "view:addr('${net}', policy.all(policy.PASS))"
+      ) fclib.networks.all}
+
+      view:addr('0.0.0.0/0', policy.all(policy.DROP))
+      view:addr('::/0', policy.all(policy.DROP))
+
+      -- load a hosts file for our assignments out of rfc1918 address space in
+      -- order to synthesise PTR records for reverse dns. don't synthesise
+      -- nodata responses for queries with mismatching address type (i.e.
+      -- resolve AAAA records using public dns for names which only have an
+      -- ipv4 address in the hosts file). this is all wrapped in some logic to
+      -- ensure that the hosts file can be reloaded by sighup at runtime.
+
+      local function load_private_hosts()
+          hints.add_hosts('/etc/nixos/rfc1918-hosts')
+          hints.use_nodata(false)
+      end
+
+      load_private_hosts()
+
+      local host_reload_due = false
+      event.recurrent(10 * sec, function (ev)
+          if host_reload_due then
+               load_private_hosts()
+               host_reload_due = false
+          end
+      end)
+
+      local signal = require('posix.signal')
+      signal.signal(signal.SIGHUP, function (sig) host_reload_due = true end)
+
+
+      --[[
+        Operator's note: control forwarders by setting
+
+           flyingcircus.roles.router.dnsForwarders = lib.mkForce [ ... ];
+
+        in the NixOS configuration. Leave the list empty to disable
+        forwarding entirely.
+      ]]
+
+      ${
+        if (role.dnsForwarders != [ ]) then
+          let
+            forwarderLiteral = lib.concatMapStringsSep ", " (f: "'${f}'") role.dnsForwarders;
+          in
+          ''
+            -- forward queries to upstream resolvers
+            policy.add(policy.all(policy.FORWARD({${forwarderLiteral}})))
+          ''
+        else
+          ''
+            -- the policy module is compliant with rfc6303 by default, override
+            -- these defaults to accept queries for rfc1918 reverse zones by the
+            -- hints module
+            policy.add(
+              policy.suffix(
+                policy.PASS,
+                policy.todnames({
+                  '10.in-addr.arpa.',
```

Comment:
I guess this is work in progress, right?

Reply:
No, this is required. If we do not have forwarders configured in kresd, so it's performing iterative resolution itself, it will refuse to answer queries for the reverse zones for RFC1918 addresses due to the defaults of the policy module (see [here](https://gitlab.nic.cz/knot/knot-resolver/-/blob/master-5/modules/policy/policy.lua?ref_type=heads#L922) and [here](https://gitlab.nic.cz/knot/knot-resolver/-/blob/master-5/modules/policy/policy.lua?ref_type=heads#L1034-1040)).

These defaults are overridden by having any policy at all, which is why we don't need the list of reverse zones when we have forwarders configured.